Goal: respect requests for uid-flattening mount options by disabling Unix permissions handling in the kernel driver

Fixes: #310982

Status wrt upstream: Forwarded on 2007/05/30

Author: Unknown

Note: Part of no-longer maintained smbfs stuff

This patch is the stopgap that was implemented immediately prior to the
sarge release in response to the security issue with the kernel ignoring
uid,gid mount options when the server supported unix capabilities.  The
corresponding changelog entry was:

 samba (3.0.14a-4) unstable; urgency=high

   [...]
   * Patch smbmount to strip CAP_UNIX out of the capabilities passed to
     the kernel when uid, gid, dmask, or fmask options have been
     specified; this keeps the mount permissions from changing out from
     under the user when upgrading to a server (or to a kernel) that
     supports unix extensions.  Closes: #310982.
   [...]

This issue has since been resolved in the kernel.  The patch should not be
included upstream in Samba, and should be dropped from the Debian packages
as well just as soon as someone has time for testing it (or, y'know, as soon
as we stop shipping mount.smbfs altogether).


Index: samba-3.0.25a/source/client/smbmount.c
===================================================================
--- samba-3.0.25a.orig/source/client/smbmount.c	2007-05-26 07:46:33.884647544 +0200
+++ samba-3.0.25a/source/client/smbmount.c	2007-05-26 07:46:34.272650637 +0200
@@ -213,6 +213,10 @@
   		c->capabilities &= ~CAP_STATUS32;
 		c->force_dos_errors = True;
 	}
+	/* For now, respect requests for uid-flattening mount options
+	   by disabling Unix permissions handling in the kernel driver */
+	if (mount_uid || mount_gid || mount_fmask || mount_dmask)
+		c->capabilities &= ~CAP_UNIX;
 
 	if (!NT_STATUS_IS_OK(cli_session_setup(c, username, 
 					       password, strlen(password),
